Simplify module

nbhuiyan opened the pull request:
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Remove unused move constructor from FunctionDef

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Remove index from functions in module format

Generated module no longer has function index with any function.
The frontend compiler and the serializer no longer outputs function
indices to module. The deseriazlier handles indexing the functions
in memory. The tests have been updated to not expect indices in
module files.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

rwy7 commented on b9/include/b9/Module.hpp:
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
-        nparams{nparams},
-        nlocals{nlocals} {}
-
+        
```
The diff looks weird. You added some whitespace, I think. Run clang-format?

nbhuiyan replied:
VS code auto indents when adding new line, and does not show up in its diff viewer by default, so I missed it. Clang-format should fix that. Good catch!

rwy7 commented on b9/include/b9/Module.hpp:
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
+        
   // Function Data
   std::string name;
   uint32_t index;
```
I don't think the index in FunctionDef is used anywhere--can we just get rid of it?

rwy7 commented on test/testDisasm.cpp:
```diff
@@ -174,14 +174,6 @@ TEST(RoundTripSerializationTest, testFunctionSection) {
-                                 END_SECTION};
-  auto f2 = FunctionDef("testName", 5, i2, 1, 2);
-  functions.push_back(f2);
-  EXPECT_THROW(roundTripFunctionSection(functions), DeserializeException);
```
What was this test even doing? Purposefully getting the index wrong?

nbhuiyan replied:
Yup! The index of the second function is 5, so `readFunctionSection()` in the deserializer throws an exception due to this:
```c++
 for (uint32_t i = 0; i < functionCount; i++) {
     functions.emplace_back("", -1, std::vector<Instruction>{});
     readFunction(in, functions.back());
     if (functions[i].index != i) {
      throw DeserializeException{"Invalid index"};
     }
   }
```

rwy7 commented on b9/include/b9/Module.hpp:
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
         instructions{instructions},
         nparams{nparams},
         nlocals{nlocals} {}
```
I think we can actually get rid of this constructor too, and just rely on the implicitly defined constructors.

nbhuiyan committed:
Run clang-format

Using -style=Google

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Rely entirely on default constructors for FunctionDef

The constructors were mainly defined to set default values for
nparams and nlocals members, which is not necessary.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan committed:
Remove index member of FunctionDef struct

FunctionDef's index was not really serving any useful purpose.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 I have addressed your comments. Also, I ran clang format on a number of files that do not have anything to do with this PR.

rwy7 commented on b9/include/b9/deserialize.hpp:
```diff
@@ -48,9 +48,10 @@ void readStringSection(std::istream &in, std::vector<std::string> &strings);
 
-void readFunctionData(std::istream &in, FunctionDef &functionSpec);
+void readFunctionData(std::istream &in, FunctionDef &functionSpec,
+                      uint32_t index);
```
I think the index can go from this parameter list.

rwy7 commented on b9/include/b9/deserialize.hpp:
```diff
@@ -48,9 +48,10 @@ void readStringSection(std::istream &in, std::vector<std::string> &strings);
+                      uint32_t index);
 
-void readFunction(std::istream &in, FunctionDef &functionDef);
+void readFunction(std::istream &in, FunctionDef &functionDef, uint32_t index);
```
same here.

rwy7 commented on b9/src/deserialize.cpp:
```diff
@@ -34,18 +34,18 @@ bool readInstructions(std::istream &in,
 
-void readFunctionData(std::istream &in, FunctionDef &functionDef) {
+void readFunctionData(std::istream &in, FunctionDef &functionDef,
+                      uint32_t index) {
```
don't need the index.

rwy7 reviewed (changes requested):
You added an index parameter to a few APIs, which we now don't need. Otherwise looks good.

nbhuiyan commented:
Totally forgot about that. That's now taken care of, thanks!
